### PR TITLE
Migrate []byte to proto Payload

### LINKED ISSUE #107: Add metadata header to all blob payload fields
Currently all blob payload fields do not have any associated metadata. It works only in monolithic applications, but might cause a lot of compatibility issues when Temporal is used as an asynchronous service mesh that connects multiple services. Another problem is that migrating applications from one encoding scheme to another is practically impossible while workflows are running. 
The strawman solution is to change all the binary blobs to `Payload` (name is up to the discussion) which includes [Header](https://github.com/temporalio/temporal-proto/blob/143d7e4a16830e2e433c019713300b5a2ad54c75/common/common.proto#L50) field.
This would allow to specify something like: _this is an encrypted field with ## version of certificate, gziped, JSON encoded._

The API proposal (Header already exists at [common.proto](https://github.com/temporalio/temporal-proto/blob/143d7e4a16830e2e433c019713300b5a2ad54c75/common/common.proto#L50))
```
message Header {
    map<string, bytes> fields = 1;
}

message Payload {
    Header header = 1;
    bytes data = 2;
}
```
Then, for example,`ActivityTaskCompletedEventAttributes` change from:
```
message ActivityTaskCompletedEventAttributes {
    bytes result = 1;
    int64 scheduledEventId = 2;
    int64 startedEventId = 3;
    string identity = 4;
}
```
to
```
message ActivityTaskCompletedEventAttributes {
    Payload result = 1;
    int64 scheduledEventId = 2;
    int64 startedEventId = 3;
    string identity = 4;
}
```
A possible alternative that potentially reduces the size of the Payload is to use an enum instead of string for metadata key names. But having that the size of payloads measures in hundreds of kilobytes I'm not sure if it is worth the complexity.

### PR DESCRIPTION
**Pull request by alexshtin**
**What changed?**
All user payload in proto files (mainly `details`, `input`, and `result` fields) was replaced from using `[]byte` to `commonpb.Payload`.

**Why?**
`commonpb.Payload` has extra metadata attached to actual data. It closes #107.

**How did you test it?**
All automatic tests including go-sdk integration tests.

**Potential risks**
User data (such activity arguments and return types) will be serialized in a wrong way.

### CODE REVIEW

File: common/codec/payload.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import (
+	commonpb "go.temporal.io/temporal-proto/common"
+	"go.temporal.io/temporal/encoded"
+)
+
+var (
+	dataConverter = encoded.GetDefaultDataConverter()
+)
+
+func EncodeString(str string) *commonpb.Payload {
```
Comment (alexshtin) [inline common/codec/payload.go:36]: This is just some helpers because they are used in many places.

Comment (shawnhathaway) [reply common/codec/payload.go:36]: Codec isnt a great name here as it isn't clear what your encoding/decoding from usage? Can we move this to its own package or call this `EncodeStringToPayload`?

File: common/codec/payload.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package codec
```
Comment (shawnhathaway) [inline common/codec/payload.go:25]: Thoughts on serialization code moving to temporal-proto-go? Starting to notice a bit of duplication (see Cadence merge as well for examples of this)

File: common/codec/payload.go (hunk 1 above)
Comment (shawnhathaway) [reply common/codec/payload.go:36]: Can we change package name to `payloadcodec` or change this to `EncodeStringToPayload`

Comment (alexshtin) [reply common/codec/payload.go]: Moved to `payload`.

File: common/codec/payload.go (hunk 2 above)
Comment (alexshtin) [reply common/codec/payload.go]: Yes, we need to do something around it. Let's discuss it one day.

**Pull request merged**